### PR TITLE
fix: added e-Waybill Status field

### PR DESCRIPTION
**Pull request by asu1609**
This field will display the status of e-way bill. It will be displayed after the sales invoice is submitted.

### CODE REVIEW

File: india_compliance/gst_india/constants/custom_fields.py [hunk 1]
```diff
@@ -798,6 +798,16 @@
         "default": "Today",
         "print_hide": 1,
     },
+    {
+        "fieldname": "ewaybill_status",
```
Comment (sagarvora) [inline india_compliance/gst_india/constants/custom_fields.py]: `e_waybill_status`

File: india_compliance/gst_india/constants/custom_fields.py [hunk 2]
```diff
@@ -798,6 +798,16 @@
...
+        "label": "e-Way Bill Status",
```
Comment (sagarvora) [inline india_compliance/gst_india/constants/custom_fields.py]: e-Waybill Status

File: india_compliance/gst_india/constants/custom_fields.py [hunk 3]
```diff
@@ -798,6 +798,16 @@
...
+        "fieldtype": "Select",
+        "insert_after": "ewaybill",
+        "options": "\nPending\nGenerated\nCancelled\nNot Applicable",
+        "default": None,
```
Comment (sagarvora) [inline india_compliance/gst_india/constants/custom_fields.py]: remove

File: india_compliance/gst_india/constants/custom_fields.py [hunk 4]
```diff
@@ -814,6 +814,18 @@
...
         "print_hide": 1,
     },
+    {
+        "fieldname": "e_waybill_status",
+        "label": "e-Waybill Status",
+        "fieldtype": "Select",
+        "insert_after": "ewaybill",
+        "options": "\nPending\nGenerated\nCancelled\nNot Applicable",
+        "print_hide": 1,
+        "read_only": 1,
```
Comment (vorasmit) [inline india_compliance/gst_india/constants/custom_fields.py]: editable. banner to depend on this.

**Comment by vorasmit** [comment]
- [ ] Manual generation and cancellation of e-Waybill flow using dialog (handle logs)

File: india_compliance/gst_india/overrides/sales_invoice.py [hunk 5]
```diff
@@ -243,3 +245,25 @@ def update_dashboard_with_gst_logs(doctype, data, *log_doctypes):
     transactions.insert(2, {"label": _("GST Logs"), "items": log_doctypes})
 
     return data
+
+
+def set_e_waybill_status(doc, method=None, gst_settings=None):
+    if doc.docstatus != 1:
+        return
+
+    if doc.ewaybill:
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/sales_invoice.py]: Only for pending vs Not Applicable

File: india_compliance/gst_india/constants/custom_fields.py [hunk 6]
```diff
@@ -814,6 +814,18 @@
...
+        "fieldname": "e_waybill_status",
+        "label": "e-Waybill Status",
+        "fieldtype": "Select",
+        "insert_after": "ewaybill",
+        "options": "\nPending\nGenerated\nCancelled\nNot Applicable",
+        "print_hide": 1,
+        "no_copy": 1,
+        "translatable": 1,
+        "allow_on_submit": 1,
+        "depends_on": "eval:doc.docstatus === 1",
```
Comment (vorasmit) [inline india_compliance/gst_india/constants/custom_fields.py]: - allow editing only for pending or not applicable.
- visible: pending or not applicable or e-Waybill is not set

File: india_compliance/hooks.py [hunk 7]
```diff
@@ -132,6 +132,7 @@
         "on_update_after_submit": (
             "india_compliance.gst_india.overrides.sales_invoice.on_update_after_submit"
         ),
+        "before_update_after_submit": "india_compliance.gst_india.overrides.sales_invoice.set_e_waybill_status",
```
Comment (vorasmit) [inline india_compliance/hooks.py]: Not handling manual. Hence not required.

File: india_compliance/patches/post_install/update_e_waybill_status_field.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
+import frappe
+from frappe.utils import add_days, nowdate
+
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_e_waybill_status_field.py]: perf.

File: india_compliance/patches/post_install/update_e_waybill_status_field.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    ).set(sales_invoice.e_waybill_status, "Pending").where(
+        ((sales_invoice.ewaybill == "") | (sales_invoice.ewaybill.isnull()))
+        & (sales_invoice.e_waybill_status.isnull())
+        & (sales_invoice.docstatus == 1)
+        & (sales_invoice.posting_date >= old_invoice_date)
+        & (sales_invoice.base_grand_total >= e_waybill_threshold)
+        & (
+            (sales_invoice_item.gst_hsn_code != "")
+            | (sales_invoice_item.gst_hsn_code.notnull())
+            | (sales_invoice_item.gst_hsn_code.not_like("99%"))
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_e_waybill_status_field.py]: non-gst

File: india_compliance/patches/post_install/update_e_waybill_status_field.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+def execute():
+    # check if e-waybill is enabled
+    if frappe.flags.in_install:
+        e_waybill_enabled = frappe.db.get_value(
+            "Sales Invoice", {"ewaybill": ["is", "set"]}
+        )
+
+    else:
+        e_waybill_enabled = frappe.db.get_single_value(
```
Comment (vorasmit) [inline india_compliance/patches/post_install/update_e_waybill_status_field.py]: only for pending

File: india_compliance/gst_india/constants/custom_fields.py [hunk 11]
```diff
@@ -814,6 +814,18 @@
...
+        "fieldname": "e_waybill_status",
+        "label": "e-Waybill Status",
+        "fieldtype": "Select",
+        "insert_after": "ewaybill",
+        "options": "\nPending\nGenerated\nCancelled\nNot Applicable",
+        "print_hide": 1,
+        "no_copy": 1,
+        "translatable": 1,
+        "allow_on_submit": 1,
+        "depends_on": "eval:doc.docstatus == 1 && !doc.ewaybill || in_list(['','Pending', 'Not Applicable'], doc.e_waybill_status)",
```
Comment (vorasmit) [inline india_compliance/gst_india/constants/custom_fields.py]: readonly depends on

**Review by vorasmit — CHANGES_REQUESTED** [review]

**Pull request merged**